modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,208 @@
+ * @access private
+ */
+function ilo_maybe_add_template_output_buffer_filter() {
+	if ( ! ilo_can_optimize_response() ) {
+		return;
+	}
+	add_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' );
+}
+add_action( 'wp', 'ilo_maybe_add_template_output_buffer_filter' );
+
+/**
+ * Constructs preload links.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param array $lcp_images_by_minimum_viewport_widths LCP images keyed by minimum viewport width, amended with attributes key for the IMG attributes.
+ * @return string Markup for zero or more preload link tags.
+ */
+function ilo_construct_preload_links( array $lcp_images_by_minimum_viewport_widths ): string {
+	$preload_links = array();
+
+	$minimum_viewport_widths = array_keys( $lcp_images_by_minimum_viewport_widths );
+	for ( $i = 0, $len = count( $minimum_viewport_widths ); $i < $len; $i++ ) {
+		$lcp_element = $lcp_images_by_minimum_viewport_widths[ $minimum_viewport_widths[ $i ] ];
+		if ( false === $lcp_element || empty( $lcp_element['attributes'] ) ) {
+			// No LCP element at this breakpoint, so nothing to preload.
+			continue;
+		}
+
+		$img_attributes = $lcp_element['attributes'];
+
+		// Prevent preloading src for browsers that don't support imagesrcset on the link element.
+		if ( isset( $img_attributes['src'], $img_attributes['srcset'] ) ) {
+			unset( $img_attributes['src'] );
+		}
+
+		// Add media query if it's going to be something other than just `min-width: 0px`.
+		$minimum_viewport_width = $minimum_viewport_widths[ $i ];
+		$maximum_viewport_width = isset( $minimum_viewport_widths[ $i + 1 ] ) ? $minimum_viewport_widths[ $i + 1 ] - 1 : null;
+		if ( $minimum_viewport_width > 0 || null !== $maximum_viewport_width ) {
+			$media_query = sprintf( '( min-width: %dpx )', $minimum_viewport_width );
+			if ( null !== $maximum_viewport_width ) {
+				$media_query .= sprintf( ' and ( max-width: %dpx )', $maximum_viewport_width );
+			}
+			$img_attributes['media'] = $media_query;
+		}
+
+		// Construct preload link.
+		$link_tag = '<link data-ilo-added-tag rel="preload" fetchpriority="high" as="image"';
+		foreach ( array_filter( $img_attributes ) as $name => $value ) {
+			// Map img attribute name to link attribute name.
+			if ( 'srcset' === $name || 'sizes' === $name ) {
+				$name = 'image' . $name;
+			} elseif ( 'src' === $name ) {
+				$name = 'href';
+			}
+
+			$link_tag .= sprintf( ' %s="%s"', $name, esc_attr( $value ) );
+		}
+		$link_tag .= ">\n";
+
+		$preload_links[] = $link_tag;
+	}
+
+	return implode( '', $preload_links );
+}
+
+/**
+ * Constructs a breadcrumbs string from a breadcrumbs array.
+ *
+ * @param array<array{tag: string, index: int}> $breadcrumbs Breadcrumbs.
+ * @return string Breadcrumb string.
+ */
+function ilo_construct_breadcrumbs_string( array $breadcrumbs ): string {
+	return implode(
+		' ',
+		array_map(
+			static function ( $breadcrumb ) {
+				return sprintf( '%s,%s', $breadcrumb['tag'], $breadcrumb['index'] );
```

Comment:
In another PR I'm working on, I'm going to make this into an XPath constructor.

```suggestion
function ilo_construct_breadcrumbs_xpath( array $breadcrumbs ): string {
	return implode(
		'',
		array_map(
			static function ( $breadcrumb ) {
				return sprintf( '/%s[%d]', $breadcrumb['tag'], $breadcrumb['index'] );
```

Reply:
See https://github.com/WordPress/performance/pull/892